src/Chainweb/Pact/Service/PactQueue.hs

```diff
@@ -51,11 +103,102 @@ addRequest q msg =  atomically $ writeTBQueue priority msg
 
 -- | Get the next available request from the Pact execution queue
 getNextRequest :: PactQueue -> IO RequestMsg
-getNextRequest q = atomically $
+getNextRequest q = do
+  (T2 req entranceTime) <- atomically $
     tryReadTBQueueOrRetry (_pactQueueValidateBlock q)
     <|> tryReadTBQueueOrRetry (_pactQueueNewBlock q)
     <|> tryReadTBQueueOrRetry (_pactQueueOtherMsg q)
+  exitTime <- getCurrentTimeIntegral
+  let requestTime = exitTime `diff` entranceTime
+      stats = case req of
+        ValidateBlockMsg {} -> _pactQueuePactQueueValidateBlockMsgStats q
+        NewBlockMsg {} -> _pactQueuePactQueueNewBlockMsgStats q
+        _ -> _pactQueuePactQueueOtherMsgStats q
+  updatePactQueueStats stats requestTime
+  return req
   where
     tryReadTBQueueOrRetry = tryReadTBQueue >=> \case
       Nothing -> retry
       Just msg -> return msg
+
+
+data PactQueueStats = PactQueueStats
+  {
+    _pactQueueStatsQueueName :: !T.Text
+  , _pactQueueStatsCounters  :: !(IORef PactQueueCounters)
+  }
+
+data PactQueueCounters = PactQueueCounters
+  {
+    _pactQueueCountersCount :: {-# UNPACK #-} !Int
+  , _pactQueueCountersSum   :: {-# UNPACK #-} !Micros
+  , _pactQueueCountersMin   :: {-# UNPACK #-} !Micros
+  , _pactQueueCountersMax   :: {-# UNPACK #-} !Micros
+  } deriving (Show, Generic)
+    deriving anyclass NFData
+
+
+instance ToJSON PactQueueCounters where
+  toJSON (PactQueueCounters {..}) =
+    object
+      [
+        "count" .= _pactQueueCountersCount
+      , "sum"   .= _pactQueueCountersSum
+      , "min"   .= _pactQueueCountersMin
+      , "max"   .= _pactQueueCountersMax
+      , "avg"   .= avg
+      ]
+    where
+      avg :: Maybe Double
+      avg = if _pactQueueCountersCount == 0 then Nothing
+        else Just $ fromIntegral _pactQueueCountersSum / fromIntegral _pactQueueCountersCount
+
+updatePactQueueStats :: PactQueueStats -> TimeSpan Micros -> IO ()
+updatePactQueueStats stats (timeSpanToMicros -> timespan) = do
+    atomicModifyIORef' (_pactQueueStatsCounters stats) $ \ctrs ->
+      (PactQueueCounters
+      {
+          _pactQueueCountersCount = _pactQueueCountersCount ctrs + 1
+        , _pactQueueCountersSum = _pactQueueCountersSum ctrs + timespan
+        , _pactQueueCountersMin = _pactQueueCountersMin ctrs `min` timespan
+        , _pactQueueCountersMax = _pactQueueCountersMax ctrs `max` timespan
+      }
+      , ())
+
+resetPactQueueStats :: PactQueue -> IO ()
+resetPactQueueStats q = do
+  resetPactQueueStats' (_pactQueuePactQueueValidateBlockMsgStats q)
+  resetPactQueueStats' (_pactQueuePactQueueNewBlockMsgStats q)
+  resetPactQueueStats' (_pactQueuePactQueueOtherMsgStats q)
+
+resetPactQueueStats' :: PactQueueStats -> IO ()
+resetPactQueueStats' stats = atomicModifyIORef' (_pactQueueStatsCounters stats) (const (initPactQueueCounters, ()))
```

Comment:
`atomicModifyIORef r (const (x, ()))` is the same as `atomicWriteIORef r x`.

Reply:
As discussed offline, it's the same except that it's not strict in `x`, which is fine here.